File: src/lexer.py
import json

PUNCTUATION = [ '+', '-', '*', '/', '%', '=', ',', '(', ')' ]
KEYWORDS = [ 'var' ]

class Token:
    def __init__(self, type, value, line):
        self.type = type
        self.value = value
        self.line = line

class Lexer:
    def __init__(self, koral_file, token_file):
        self.koral_file = koral_file
        self.token_file = token_file
        self.lexemes = []
        self.tokens = []
        self.lines = []

    def execute(self):
        self.scan(self.koral_file)
        self.evaluate(self.lexemes, self.tokens)
        self.generate_token_base(self.tokens, self.token_file)

    def scan(self, file):
        with open(file, 'r') as f:
            in_string = False
            lexeme_count = len(self.lexemes)
            for index, line in enumerate(f.readlines()):
                current = ''
                for i in range(len(line)):
                    if line[i] == '"':
                        current += line[i]
                        if in_string:
                            self.lexemes.append(current)
                            current = ''
                        in_string = not in_string
                    elif line[i] in PUNCTUATION and not in_string:
                        if current != '':
                            self.lexemes.append(current)
                        self.lexemes.append(line[i])
                        current = ''
                    elif line[i] != ' ' or in_string:
                        current += line[i]
                    if current in KEYWORDS:
                        self.lexemes.append(current)
                        current = ''
                    if i == len(line) - 1 and (in_string or current != ''):
                        self.lexemes.append(current)
                    if line[i] == ' ' and not in_string:
                        if current != '':
                            self.lexemes.append(current)
                        current = ''
                    for i in range(len(self.lexemes) - lexeme_count):
                        self.lines.append(index + 1)
                    lexeme_count = len(self.lexemes)

    def process_unary(self, tokens):
        for index, token in enumerate(tokens):
            if tokens[index-1].type == 'Operator' and token.type == 'Operator' and tokens[index+1].type == 'Numeric':
                token.value += tokens[index+1].value
                token.type = 'Numeric'
                tokens.pop(index+1)

    def determine_type(self, lexeme):
        if lexeme in KEYWORDS:
            return 'Keyword'

        type = {
            '+': 'Operator',
            '-': 'Operator',
            '*': 'Operator',
            '/': 'Operator',
            '%': 'Operator',
            '=': 'Operator',
            ',': 'Punctuator',
            '(': 'Punctuator',
            ')': 'Punctuator'
        }.get(lexeme, 'Identifier')

        if type == 'Identifier':
            if lexeme[0] == '\'' or lexeme[0] == '\"':
                return 'String'
            try:
                float(lexeme)
            except:
                return type
            return 'Numeric'

        return type

    def evaluate(self, lexemes, tokens):
        for index, lexeme in enumerate(lexemes):
            if lexeme == '\n':
                continue
            
            type = self.determine_type(lexeme)
            tokens.append(Token(type, lexeme, self.lines[index]))
        self.process_unary(tokens)

    def generate_token_base(self, tokens, token_base):
        token_list = []
        with open(token_base, 'w') as f:
            for token in tokens:
                token_list.append({"type": token.type, "value": token.value})
            json.dump(token_list, f, indent=4)

File: src/Lexer.cpp
#include "Lexer.hpp"

Token::Token(std::string value, size_t line)
{
    this->value = value;
    this->line = line;
}

std::vector<std::shared_ptr<Token>> Lexer::Lex()
{
    Scan();
    Evaluate();

    PostProcess();

    return tokens;
}

void Lexer::Scan()
{
    std::vector<std::string> lines = ReadLines();
    bool inString = false;

    for (size_t i = 0; i < lines.size(); i++)
    {
        std::string line = lines[i];
        std::string lexeme = "";

        for (auto character : line)
        {
            ProccessCharacter(character, lexeme, inString, i);
        }

        if (lexeme != "")
        {
            AddToken(lexeme, i);
        }
    }
}

void Lexer::ProccessCharacter(char character, std::string& lexeme, bool& inString, size_t line)
{
    if (character == '\"' || character == '\'')
    {
        inString = !inString;
    }

    if (inString)
    {
        lexeme += character;
        return;
    }

    if (character == ' ')
    {
        if (lexeme != "")
        {
            AddToken(lexeme, line);
        }

        return;
    }
    
    if (std::find(punctuation.begin(), punctuation.end(), character) != punctuation.end() ||
        std::find(operators.begin(), operators.end(), character) != operators.end())
    {
        if (lexeme != "")
        {
            AddToken(lexeme, line);
        }

        lexeme += character;
        AddToken(lexeme, line);

        return;
    }

    lexeme += character;
}

void Lexer::Evaluate()
{
    for (auto& token : tokens)
    {
        std::string type = DetermineType(token->GetValue());
        token->SetType(type);
    }
}

void Lexer::PostProcess()
{
    for (size_t i = 1; i < tokens.size() - 1; i++)
    {
        auto& previousToken = tokens[i - 1];
        auto& token = tokens[i];
        auto& nextToken = tokens[i + 1];

        std::vector<std::shared_ptr<Token>> tokensToCheck = { previousToken, token, nextToken };

        std::vector<std::string> unary = { "Operator", "Operator", "Number" };
        std::vector<std::string> floating = { "Number", "Punctuation", "Number" };

        if (AreForwarded(tokensToCheck, unary))
        {
            if (token->GetValue() == "-" || token->GetValue() == "+")
            {
                std::string value = token->GetValue() + nextToken->GetValue();
                nextToken->SetValue(value);

                tokens.erase(tokens.begin() + i);
            }
        }

        if (AreForwarded(tokensToCheck, floating))
        {
            if (token->GetValue() == ".")
            {
                std::string value = previousToken->GetValue() + token->GetValue() + nextToken->GetValue();
                nextToken->SetValue(value);

                tokens.erase(tokens.begin() + i);
                tokens.erase(tokens.begin() + i - 1);
            }
        }
    }
}

std::string Lexer::DetermineType(std::string lexeme)
{
    if (std::find(keywords.begin(), keywords.end(), lexeme) != keywords.end())
        return "Keyword";

    if (std::find(types.begin(), types.end(), lexeme) != types.end())
        return "Type";

    if (std::find(operators.begin(), operators.end(), lexeme[0]) != operators.end())
        return "Operator";
    
    if (std::find(punctuation.begin(), punctuation.end(), lexeme[0]) != punctuation.end())
        return "Punctuation";
    
    if (lexeme[0] == '"' || lexeme[0] == '\'')
        return "String";
    
    try
    {
        std::stoi(lexeme);
        return "Number";
    }
    catch (std::invalid_argument const&)
    {
        return "Identifier";
    }

    return "Unknown";
}

std::vector<std::string> Lexer::ReadLines()
{
    std::vector<std::string> lines;
    std::string line;

    for (size_t i = 0; i < content.length(); i++)
    {
        if (content[i] != '\n')
        {
            line += content[i];
            continue;
        }

        lines.push_back(line);
        line = "";
    }

    lines.push_back(line);

    return lines;
}

bool Lexer::AreForwarded(std::vector<std::shared_ptr<Token>> tokens, std::vector<std::string> expected)
{
    if (tokens.size() != expected.size())
    {
        return false;
    }
    
    for (size_t i = 0; i < tokens.size(); i++)
    {
        if (tokens[i]->GetType() != expected[i])
        {
            return false;
        }
    }

    return true;
}

void Lexer::AddToken(std::string& lexeme, size_t line)
{
    auto token = std::make_shared<Token>(lexeme, line);
    
    tokens.push_back(token);

    lexeme = "";
}